File: samples/graph/graph.js
/*

  Graph support functions

  This code all belongs in `defjs` functions, but I haven't
  written the compiler for that yet.

*/

displaylist = [];

paintchart = function () {
    cv = document.getElementById('cv');
    ctx = cv.getContext("2d");
    ctx.fillStyle = 'white';
    ctx.fillRect(0, 0, cv.width, cv.height);
    ctx.save();
    ctx.scale(cv.width/13, cv.height/3);
    ctx.clearRect(0, 0, 13, 3);
    ctx.translate(0, 1.5);
    ctx.beginPath();
    ctx.moveTo(0, 0);
    ctx.lineTo(13, 0);
    ctx.moveTo(0, 0);
    displaylist.forEach((op) => op(ctx));
    ctx.restore();
    ctx.stroke();
}

menu = function (msg) {
    webSocket.send(msg.target.id);
}

File: samples/tab/graph.c
#include <stdio.h>

void
beginpaint()
{
	printf("dl = new Array();\n");
}

void
endpaint()
{
	printf("dl.forEach((op) => op(cx));\n");
}

void
beginpath()
{
	printf("dl.push((cx) => cx.beginPath());\n");
}

void
clearrect(int x, int y, int width, int height)
{
	printf("dl.push((cx) => cx.clearRect(%d, %d, %d, %d));\n", x, y, width, height);
}

void
filltext(const char *s, int x, int y)
{
	printf("dl.push((cx) => cx.fillText('%s', %d, %d));\n", s, x, y);
}

void
lineto(int x, int y)
{
	printf("dl.push((cx) => cx.lineTo(%d, %d));\n", x, y);
}

void
moveto(int x, int y)
{
	printf("dl.push((cx) => cx.moveTo(%d, %d));\n", x, y);
}

void
restore()
{
	printf("dl.push((cx) => cx.restore());\n");
}

void
save()
{
	printf("dl.push((cx) => cx.save());\n");
}

void
setlinewidth(double x)
{
	printf("dl.push((cx) => cx.lineWidth = %f);\n", x);
}

void
setstrokestyle(const char *s)
{
	printf("dl.push((cx) => cx.strokeStyle = '%s');\n", s);
}

void
stroke()
{
	printf("dl.push((cx) => cx.stroke());\n");
}

void
strokerect(int x, int y, int width, int height)
{
	printf("dl.push((cx) => cx.strokeRect(%d, %d, %d, %d));\n", x, y, width, height);
}

void
translate(int x, int y)
{
	printf("dl.push((cx) => cx.translate(%d, %d));\n", x, y);
}



/*
void show(int i, int x, int y)
{
	printf("%s.fillText('%d', %d - %s.measureText('%d').width / 2, %d)\n",
		cx, fret[i], x, cx, fret[i], y);
}
*/


